### PR TITLE
Improve performance of updating interruption by 1 order of magnitude

### PR DESCRIPTION
**Pull request by kyri-petrou**
A very large number of ZIO methods require updating the interruption status of an effect (`onExit`, `ensuring`, `acquireRelease*`, `uninterruptibleMask`, etc.)

When running benchmarks recently, I noticed that the performance of such methods was quite poor, and the evaluation of `UpdateRuntimeFlagsWithin` was significantly slower than other effects. The reasons AFACT for this is are:
1. type-checkin on an interface is significantly slower than type-checking on a concrete class (within `FiberRuntime`)
2. Interface method calls are slower than concrete method calls.
3. We're checking the interruption status of the fiber multiple times within a single update of the interruption status

So this PR optimizes these through the following changes:
1. Used `UpdateRuntimeFlagsWithin.DynamicNoBox` as the only class that updates runtime flags. In a perfect world where we didn't have to care about binary compatibility, we could just delete the  `UpdateRuntimeFlagsWithin` trait and make it a case class, but this will have to do for now.
2. Changed the logic in `FiberRuntime` to handle `DynamicNoBox` in pretty much the same way that we were handling `UpdateRuntimeFlagsWithin` (Note: I added an unreachable case statement to ensure backwards compatibility _just in case_
3. Added the `_isInterrupted` var within FiberRuntime which we can check for interruption. This is much faster than checking the FiberRefs Map

### Benchmarks:

TLDR: we're looking at an 8-10x speedup, with `ZIO.uninterruptibleMask` having the biggest improvement (~x10!!) 🚀 

series/2.x:
```
[info] Benchmark                               (nesting)   Mode  Cnt        Score       Error  Units
[info] RegionBenchmark.zioBracket                    100  thrpt    4   107032.744 ±  2107.894  ops/s
[info] RegionBenchmark.zioEnsuring                   100  thrpt    4   140833.271 ±  1918.259  ops/s
[info] RegionBenchmark.zioUninterruptible            100  thrpt    4  5341635.192 ± 99473.848  ops/s
[info] RegionBenchmark.zioUninterruptible2           100  thrpt    4   257882.824 ±  4007.369  ops/s
[info] RegionBenchmark.zioUninterruptibleMask        100  thrpt    4   122529.491 ± 36266.328  ops/s
```

PR:
```
[info] Benchmark                               (nesting)   Mode  Cnt        Score         Error  Units
[info] RegionBenchmark.zioBracket                    100  thrpt    4   830526.258 ±   24216.089  ops/s
[info] RegionBenchmark.zioEnsuring                   100  thrpt    4   924147.162 ±   22944.079  ops/s
[info] RegionBenchmark.zioUninterruptible            100  thrpt    4  4850824.796 ± 1941039.646  ops/s
[info] RegionBenchmark.zioUninterruptible2           100  thrpt    4  2493429.718 ±   76481.962  ops/s
[info] RegionBenchmark.zioUninterruptibleMask        100  thrpt    4  1253910.119 ±   10401.410  ops/s
```

Note on the `RegionBenchmark.zioUninterruptible` case that is a bit slower: This became slower because of the allocation of the `IntFunction` which we previously didn't do. However, this benchmark is highly unrealistic because:
1. The benchmark code path results in the C2 compiler seeing a single implementation of `UpdateRuntimeFlagsWithin` (Uninterruptible). Therefore, it makes an optimistic assumption that this will be the only implementation, and so it JIT compiles the type-checking on `UpdateRuntimeFlagsWithin`
2. The `uninterruptible` method is used recursively, which means the interruption flag is not really updated

### CODE REVIEW

**Comment by kyri-petrou** [comment]
@jdegoes sorry to bug you, but would you be able to review this PR or shall I try and find someone else to review it? Unfortunately not many people are confident in reviewing PRs that touch the internals of `FiberRuntime` 🥲

**Comment by calvinlfer** [comment]
Can we create an issue for this? 
This is bounty worthy 🙏

**Comment by kyri-petrou** [comment]
@calvinlfer I've never seen this used before, but _I think_ it's possible to add a tip to a PR directly without an issue/bounty

https://docs.algora.io/tipping

**Comment by kyri-petrou** [comment]
@calvinlfer let me know if that doesn't work or if you prefer to open an issue for this instead and I can do that

**Review by ghostdogpr — APPROVED** [review]
